### PR TITLE
Support wildcards in valid_prog_environs

### LINKED ISSUE #224: Support wildcards in 'valid_prog_environs'
It would be useful to specify programming environments like that:

```python
self.valid_prog_environs = ['PrgEnv-gnu-*', 'foss/*']
```

### PR DESCRIPTION
**Pull request by teojgo**
* Add support for wildcards in `valid_prog_environs` of a `RegressionTest`.
* Create the corresponding unit tests.
* Update doc for the `valid_prog_environs` field.

Closes #224

### CODE REVIEW

File: reframe/core/pipeline.py [hunk 1]
```diff
@@ -676,10 +681,10 @@ def supports_system(self, partition_name):
         return partition_name in self.valid_systems
 
     def supports_environ(self, env_name):
-        if '*' in self.valid_prog_environs:
-            return True
-
-        return env_name in self.valid_prog_environs
+        for env in self.valid_prog_environs:
+            if fnmatch.fnmatch(env_name, env):
+                return True
```
Comment (vkarak) [inline reframe/core/pipeline.py]: Please leave a blank newline here.

File: unittests/test_pipeline.py [hunk 2]
```diff
@@ -242,6 +242,27 @@ def test_supports_system(self):
...
+
+        test.valid_prog_environs = ['PrgEnv-foo-*']
+        self.assertTrue(test.supports_environ('PrgEnv-foo-version1'))
+        self.assertTrue(test.supports_environ('PrgEnv-foo-version2'))
+        self.assertFalse(test.supports_environ('PrgEnv-boo-version1'))
+
+        test.valid_prog_environs = ['PrgEnv-*']
+        self.assertTrue(test.supports_environ('PrgEnv-foo1'))
+        self.assertTrue(test.supports_environ('PrgEnv-foo2'))
+        self.assertFalse(test.supports_environ('Prgenv-foo'))
```
Comment (vkarak) [inline unittests/test_pipeline.py]: I don't see the reason for this set of checks. They are equivalent to the above block. You can simply add a check for case sensitivity.

File: reframe/core/pipeline.py [hunk 3]
```diff
@@ -676,10 +681,10 @@ def supports_system(self, partition_name):
         return partition_name in self.valid_systems
 
     def supports_environ(self, env_name):
-        if '*' in self.valid_prog_environs:
-            return True
-
-        return env_name in self.valid_prog_environs
+        for env in self.valid_prog_environs:
+            if fnmatch.fnmatch(env_name, env):
```
Comment (vkarak) [inline reframe/core/pipeline.py]: I like this and it's simple. Would it make sense to support wildcards also in the `supports_system()`?

Comment (teojgo) [reply reframe/core/pipeline.py]: It is possible to do so if you want to support also wildcards for systems.

**Pull request merged**